### PR TITLE
feat_: recalculate route fees with every new block

### PR DESCRIPTION
**Pull request by saledjenic**
Base branch is `router-clean-up`, please review it first. :)

Since we cannot use `SubscribeNewHead` call to be notified about every new block creation, cause our providers do not support that approach, we have to check for a new block creation in some intervals.

They are set here:
https://github.com/status-im/status-go/blob/6c3f87b5e18fabe697f485923ac574336a81b378/services/wallet/router/router_updates.go#L12-L18

Every time fees get recalculated a signal `"wallet.suggested.routes"` will be sent, and the uuid will be the same as when the client asked for routes. 

If the client calculating route was initiated calling `GetSuggestedRoutesAsync` or `GetSuggestedRoutes` then it should be closed calling `StopSuggestedRoutesAsyncCalcualtion` or `StopSuggestedRoutesCalcualtion`.

I've tested this in the desktop app, and it works nicely. Will open a PR for the desktop app later.

### CODE REVIEW

**Comment by saledjenic** [comment]
If you think it's good, maybe we can add some timeout that will stop sending updates to a client after 5-10mis, in case somebody opens the send flow and keeps it like that for hours? 

Happy to add/change anything, expecting your suggestions. :)

File: services/wallet/router/router_updates.go
Comment (alaibe) [inline services/wallet/router/router_updates.go]: every 2 milliseconds?

**Comment by alaibe** [comment]
@saledjenic make sure to merge without mobile changes too to avoid not canceling the fees update.

I think the fees should stop being generated automatically after x minute (a timeout in the polling)
I'd vote for 5min as it is more than enough IMO

File: services/wallet/router/router_updates.go
Comment (alaibe) [reply services/wallet/router/router_updates.go]: I'd suggest every 6 seconds for mainnet. we will be a bit off but far from wrong/

Also what happens if in between u don't have enought to pay?

Comment (saledjenic) [reply services/wallet/router/router_updates.go]: Timeout set to 5 minutes, we're checking every 3 seconds for a new block on mainnet, if needed we will adjust that later (it's a single line change).

Comment (briansztamfater) [reply services/wallet/router/router_updates.go]: For swaps, for example, we need to recalculate fees every 15 seconds https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=124-19384&t=UMFqL7n7Hgp00nTo-4

**Comment by briansztamfater** [comment]
Would be great to also send a signal right before we start to recalculate fees, so we can properly update UI in clients with loading states

**Comment by saledjenic** [comment]
> Would be great to also send a signal also right before we start to recalculate fees, so we can properly update UI in clients with loading states

@briansztamfater are you sure that's really needed?
Depending on the chains (route, which is a list of paths) that participate in the best route we update fees in frequency defined for that chain (mentioned in the description).

How it should be used and how it works:
- user runs the send flow
- client app asks statusgo to generate the best route calling `GetSuggestedRoutesAsync` - at this point the app should display loading states, while waiting for the best route, once the best route is received we display it
- after `GetSuggestedRoutesAsync` call generates the best route it subscribes for fee updates and from that moment on, fees are calculated at new block creation intervals and signal is sent to a client
- the client should display fees received from the previous response all the time until it gets new fees, at that moment client should just replace fees, without any loading states, no need for that.

**Comment by saledjenic** [comment]
In the second commit of this PR, I've added checks for balances after the fees update, so the initially generated route remains the same, we're updating fees and checking if the selected account has enough balance for that route only.

File: services/wallet/router/router.go
Comment (dlipicar) [inline services/wallet/router/router.go:124]: we really need to stop doing this sort of thing in favor of mocked dependencies...

Comment (saledjenic) [reply services/wallet/router/router.go:124]: That's not done here, that is the same thing that was done 2-3 months ago, I just had to create that function cause I moved the code around.

Comment (saledjenic) [reply services/wallet/router/router.go:124]: And I agree, potentially we should get rid of that, but somehow we still haven't come to it.

File: services/wallet/router/router.go
Comment (dlipicar) [inline services/wallet/router/router.go]: typo `Calcualtion` here and the one above. I guess we can keep it at the API level for client compatibility, but no need to keep it here :D

Comment (saledjenic) [reply services/wallet/router/router.go]: @dlipicar updated in both places, better now than later. ;)

**Comment by saledjenic** [comment]
Could you let me know if you would like anything else to be updated here?

If not I am waiting for approval to merge this into the base branch, this PR https://github.com/status-im/status-go/pull/5778 and after breaking changes get handled on the mobile app side, we can merge that. 

Desktop app PRs that handle these changes are here:
- https://github.com/status-im/status-desktop/pull/16219

**Comment by briansztamfater** [comment]
> > Would be great to also send a signal also right before we start to recalculate fees, so we can properly update UI in clients with loading states
> 
> @briansztamfater are you sure that's really needed? Depending on the chains (route, which is a list of paths) that participate in the best route we update fees in frequency defined for that chain (mentioned in the description).
> 
> How it should be used and how it works:
> 
> * user runs the send flow
> * client app asks statusgo to generate the best route calling `GetSuggestedRoutesAsync` - at this point the app should display loading states, while waiting for the best route, once the best route is received we display it
> * after `GetSuggestedRoutesAsync` call generates the best route it subscribes for fee updates and from that moment on, fees are calculated at new block creation intervals and signal is sent to a client
> * the client should display fees received from the previous response all the time until it gets new fees, at that moment client should just replace fees, without any loading states, no need for that.

Well, at least is a requirement for swaps in mobile (showing loading states for recalculating fees), but we can reconfirm it with @xAlisher. As I see in Metamask, fees do some fade animation when updated, but it is different from a loading state and action buttons aren't blocked. If we implement a similar approach on mobile instead of showing loading states (like in this design https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=124-18563&t=C9xDTrwEeF0v7u3Q-4), I think this PR is good in its current state.

**Comment by saledjenic** [comment]
> but it is different from a loading state and action buttons aren't blocked

@briansztamfater yes, exactly that, the only loading state we need is while we're waiting for the route to be generated, after that when updating just a fees part of the route we don't need it. When newer fees are received we just display them, and it's a pure UI thing whether we're going to display it by simply replacing it or adding some effect like fade out/in.

**Review by alaibe — APPROVED** [review]

**Comment by briansztamfater** [comment]
@saledjenic Just checked with @xAlisher, he thinks we should disable call to action button during recalculation, so we should also need a signal to be sent before recalculating fees too. Not sure about the effort of adding this, but we can do it in another PR if we need to merge this ASAP.

**Comment by saledjenic** [comment]
Ok, I will merge this don't want to keep it in PR anymore, and if needed we can add that signal later via separate PR.

But in my opinion... on mainnet a new block is generated every 12 seconds and maybe it makes sense (but not needed), but on optimism is every 2 seconds, on arbitrum every 250ms, does that mean that "Send" button will blink all the time?

@briansztamfater @xAlisher

**Comment by xAlisher** [comment]
> Ok, I will merge this don't want to keep it in PR anymore, and if needed we can add that signal later via separate PR.
> 
> But in my opinion... on mainnet a new block is generated every 12 seconds and maybe it makes sense (but not needed), but on optimism is every 2 seconds, on arbitrum every 250ms, does that mean that "Send" button will blink all the time?
> 
> @briansztamfater @xAlisher

Agree,  2 seconds and 250ms is to often to blink the button.  Maybe we even do not need a loading state.  I need to think about it a bit more @saledjenic @briansztamfater

**Comment by saledjenic** [comment]
The point is to disable the Send button only when there is not enough balance after the fees update, otherwise, for a better visual effect we can just fade out old and fade in new fees.

This is what Metamask does.
@xAlisher

**Pull request merged**
